dirty_cat/datetime_encoder.py

```diff
@@ -0,0 +1,171 @@
+from sklearn.base import BaseEstimator, TransformerMixin
+import numpy as np
+import pandas as pd
+from dirty_cat.utils import check_input
+from pandas.tseries.holiday import USFederalHolidayCalendar as calendar
+from typing import Union, Optional, List
+
+
+
+class DatetimeEncoder(TransformerMixin, BaseEstimator):
+    """
+    This encoder transforms each datetime column into several numeric columns corresponding to temporal features,
+    e.g year, month, day...
+    Constant extracted features are dropped ; for instance, if the year is always the same in a feature, the extracted "year" column won't be added.
+    If the dates are timezone aware, all the features extracted will correspond to the provided timezone.
+
+    Parameters
+    ----------
+    extract_until : {"year", "month", "day", "hour", "minute", "second",
+    "millisecond", "microsecond", "nanosecond"}, default="hour"
+        Extract up to this granularity, and gather the rest into the "other" feature.
+        For instance, if you specify "day", only "year", "month", "day" and "other" features will be created.
+        The "other" feature will be a numerical value expressed in the "extract_until" unit.
```

Comment:
Hmm interesting. I get you point, but here's why I chose to do it this way:
- I fear that full_time_to_epoch would be very collinear with the highest-level feature, hurting interpretability. For a worst case scenario, we can imagine a dataset of different hours during the same day for different years.
- It’s true that with my choice, the learner has to learn the weird algebra. But with your proposition, the learner would still need to learn this weird algebra if it wants to use my `other` feature (which may be more important for prediction since it’s information which isn’t contained in the other variables), because `other` = full_time_to_epoch - (year - 1970) * 365 * 24 * 3600 + …
Maybe using both could indeed make the learning easier, but wouldn't you worry about collinearity?
All in all I'm not sure.

Reply:
I do machine learning. I don't worry about collinearity :).

It does not hurt for prediction. I can hurt for interpretability.

Reply:
My reason for favoring the full_time_to_epoch is that I think that, in general, it is more likely to be useful than the other end, and we are more likely to have to reconstruct it.